packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2285,7 +2287,19 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
     switch (defaultTargetPlatform) {
       case TargetPlatform.iOS:
       case TargetPlatform.macOS:
-        _caretPrototype = Rect.fromLTWH(0.0, 0.0, cursorWidth, cursorHeight + 2);
+        final double styleHeight = _textPainter.text?.style?.height ?? 1.0;
+        // Extend the cursor height only when the text style height is less than 1.3.
+        // Extending the cursor height made sense on M2 whose default text style had
+        // a 1.0 line height. On M3, whose default line height is 1.5, the extension
+        // looks bad and might overflow on the label.
+        // When line height is equal or superior to 1.3, cursor height is reduced by
+        // a small amount to get a better rendering.
+        // TODO(bleroux): replace this fix with a long term solution, see
+        // https://github.com/flutter/flutter/issues/7034 and
+        // https://github.com/flutter/flutter/issues/120836.
+        _caretPrototype = styleHeight < 1.3
+          ? Rect.fromLTWH(0.0, 0.0, cursorWidth, cursorHeight + _kCaretHeightExtention)
+          : Rect.fromLTWH(0.0, _kCaretHeightExtention / 2, cursorWidth, cursorHeight - _kCaretHeightExtention);
```

Comment:
Below, for the other platforms on line 2293/2307, should the hardcoded `- 2.0` be using _kCaretHeightExtention too?

Reply:
In this line the 2.0 is a multiplicator.
This is due to this code focusing on the offset so the extension is 2 * offset. 
I added a new constant to avoid mixing extension and offset : in the existing code `Rect.fromLTWH(0.0, 0.0, cursorWidth, cursorHeight + 2` there was no offset so it would have been weird to replace it with `Rect.fromLTWH(0.0, 0.0, cursorWidth, cursorHeight + _kCaretHeightOffset` where the offset is in fact a size extension.
This probably illustrate that a deeper refactoring/rewriting is needed for this existing code.